File: packages/evals/tasks/bench/act/iframe_scroll.ts
import { defineBenchTask } from "../../../framework/defineTask.js";

export default defineBenchTask(
  { name: "iframe_scroll" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto(
        "https://browserbase.github.io/stagehand-eval-sites/sites/iframe-same-proc-scroll/",
      );
      await stagehand.act("scroll down 50% inside the iframe");

      await new Promise((resolve) => setTimeout(resolve, 5000));

      // Measure the scroll position inside the same-origin iframe.
      const scrollInfo = await page.evaluate(() => {
        const iframe = document.querySelector("iframe");
        const win = iframe?.contentWindow;
        const doc = iframe?.contentDocument;
        if (!win || !doc) {
          throw new Error("could not access iframe content");
        }
        return {
          scrollTop: win.scrollY + win.innerHeight / 2,
          scrollHeight: doc.documentElement.scrollHeight,
        };
      });

      const halfwayScroll = scrollInfo.scrollHeight / 2;
      const halfwayReached = Math.abs(scrollInfo.scrollTop - halfwayScroll) <= 1;
      const evaluationResult = halfwayReached
        ? {
            _success: true,
            logs: logger.getLogs(),
            debugUrl,
            sessionUrl,
          }
        : {
            _success: false,
            logs: logger.getLogs(),
            debugUrl,
            sessionUrl,
            message: `Scroll position (${scrollInfo.scrollTop}px) is not halfway down the page (${halfwayScroll}px).`,
          };

      return evaluationResult;
    } catch (error) {
      return {
        _success: false,
        error: error instanceof Error ? error.message : String(error),
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } finally {
      await stagehand.close();
    }
  },
);

File: packages/sdk-ruby/evals/tasks/act/iframe_scroll.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/act/iframe_scroll.ts

Evals.define_task("iframe_scroll") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/iframe-same-proc-scroll/")
  t.stagehand.act("scroll down 50% inside the iframe")

  t.page.wait_for_timeout(5000)

  # v3 evaluated inside page.frames()[1]; v4 exposes no frames() list,
  # so the same measurement is re-expressed via the same-origin iframe's
  # contentWindow/contentDocument.
  scroll_info = t.page.evaluate(<<~JS)
    (() => {
      const iframe = document.querySelector("iframe");
      const win = iframe?.contentWindow;
      const doc = iframe?.contentDocument;
      if (!win || !doc) {
        throw new Error("could not access iframe content");
      }
      return {
        scrollTop: win.scrollY + win.innerHeight / 2,
        scrollHeight: doc.documentElement.scrollHeight,
      };
    })()
  JS

  scroll_top = scroll_info["scrollTop"]
  halfway_scroll = scroll_info["scrollHeight"] / 2.0
  halfway_reached = (scroll_top - halfway_scroll).abs <= 1

  if halfway_reached
    { _success: true }
  else
    {
      _success: false,
      message: "Scroll position (#{scroll_top}px) is not halfway down the page (#{halfway_scroll}px).",
    }
  end
end
